src/ini-load.js

```diff
@@ -56,13 +86,78 @@ export function whenContentIniLoad(
     const promises = [];
     resources.forEach((r) => {
       if (!EXCLUDE_INI_LOAD.includes(r.element.tagName)) {
-        promises.push(r.loadedOnce());
+        promises.push(r.element.whenLoaded());
       }
     });
     return Promise.all(promises);
   });
 }
 
+/**
+ * A new way using direct measurement.
+ *
+ * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+ * @param {boolean=} opt_prerenderableOnly signifies if we are in prerender mode.
+ * @return {!Promise}
+ * @visibleForTesting
+ * TODO(#31915): remove visibility
+ */
+export function whenContentIniLoadInOb(elementOrAmpDoc, opt_prerenderableOnly) {
+  const ampdoc = Services.ampdoc(elementOrAmpDoc);
+  // First, wait for the `ready-scan` signal. Waiting for each element
+  // individually is too expensive and `ready-scan` will cover most of
+  // the initially parsed elements.
+  const whenReady = ampdoc.signals().whenSignal(READY_SCAN_SIGNAL);
+  return whenReady.then(() => {
+    // Filter elements.
+    const resources = Services.resourcesForDoc(ampdoc);
+    const elements = resources
+      .get()
+      .filter((r) => {
+        if (opt_prerenderableOnly && !r.prerenderAllowed()) {
+          return false;
+        }
+        return !EXCLUDE_INI_LOAD.includes(r.element.tagName);
+      })
+      .map((r) => r.element);
+
+    if (elements.length === 0) {
+      return Promise.resolve([]);
+    }
+
+    // Find intersecting elements.
+    return new Promise((resolve) => {
```

Comment:
nit:  could be:
```js
return Promise.all(
  elements.slice(0,100)
          .map(measureIntersection)
          .filter(entry => entry.isIntersecting)
          .map(entry => entry.element.whenLoaded()
);
```

Reply:
I really didn't want to measure one-by-one. There's a bit benefit from doing it guaranteed in bulk. The API ensures that they are all returned in one response.

Reply:
I can add a unit test, but I'm pretty sure `measureIntersection` is currently built s.t. it would all happen in a single pass as well

Reply:
Yeah. Certainly it'd work correctly. But I just felt that this use case is so separate from everything we're doing that having a new InOb there actually helps.